polkadot/runtime/parachains/src/coretime/mod.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Extrinsics implementing the relay chain side of the Coretime interface.
+//!
+//! <https://github.com/polkadot-fellows/RFCs/blob/main/text/0005-coretime-interface.md>
+
+use sp_std::{prelude::*, result};
+
+use frame_support::{pallet_prelude::*, traits::Currency};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use pallet_broker::{CoreAssignment, CoreIndex as BrokerCoreIndex};
+use primitives::{CoreIndex, Id as ParaId};
+use sp_arithmetic::traits::SaturatedConversion;
+use xcm::v3::{
+	send_xcm, Instruction, Junction, Junctions, MultiLocation, OriginKind, SendXcm, Xcm,
+};
+
+use crate::{
+	assigner_coretime::{self, PartsOf57600},
+	initializer::{OnNewSession, SessionChangeNotification},
+	origin::{ensure_parachain, Origin},
+};
+
+mod benchmarking;
+pub mod migration;
+
+pub trait WeightInfo {
+	fn request_core_count() -> Weight;
+	//fn request_revenue_info_at() -> Weight;
+	//fn credit_account() -> Weight;
+	fn assign_core(s: u32) -> Weight;
+}
+
+/// A weight info that is only suitable for testing.
+pub struct TestWeightInfo;
+
+impl WeightInfo for TestWeightInfo {
+	fn request_core_count() -> Weight {
+		Weight::MAX
+	}
+	// TODO: Add real benchmarking functionality for each of these to
+	// benchmarking.rs, then uncomment here and in trait definition.
+	/*fn request_revenue_info_at() -> Weight {
+		Weight::MAX
+	}
+	fn credit_account() -> Weight {
+		Weight::MAX
+	}*/
+	fn assign_core(_s: u32) -> Weight {
+		Weight::MAX
+	}
+}
+
+/// Shorthand for the Balance type the runtime is using.
+// type BalanceOf<T> =
+//     <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// Broker pallet index on the coretime chain. Used to
+///
+/// construct remote calls. The codec index must correspond to the index of `Broker` in the
+/// `construct_runtime` of the coretime chain.
+#[derive(Encode, Decode)]
+enum BrokerRuntimePallets {
+	#[codec(index = 50)]
+	Broker(CoretimeCalls),
+}
+
+/// Call encoding for the calls needed from the Broker pallet.
+#[derive(Encode, Decode)]
+enum CoretimeCalls {
+	#[codec(index = 1)]
+	Reserve(pallet_broker::Schedule),
+	#[codec(index = 3)]
+	SetLease(pallet_broker::TaskId, pallet_broker::Timeslice),
+	#[codec(index = 19)]
+	NotifyCoreCount(u16),
+}
```

Comment:
Before we launch this on Kusama/Polkadot, these should move to the runtime level. Similar to how it is done in the broker with a trait.

Reply:
https://github.com/orgs/paritytech/projects/119/views/20?pane=issue&itemId=48208930